### ISSUE TITLE
R6.5: Dependency injection — Trainer, BentoML, data loader decoupling

### ISSUE THREAD

**Issue by petteriTeikari** [minivess-mlops/minivess-mlops#56]
## Context

The decoupling analysis identified tight coupling in the Trainer (constructs own optimizer/scheduler), BentoML service (doesn't use ModelAdapter interface), and data loader (hardcodes transforms). These limit testability and extensibility.

Ref: `docs/planning/code-review-report-v0-2-2nd-pass.md` — R5.13, R5.14, R5.17
Ref: Background decoupling agent findings

### R5.14: Trainer dependency injection (~4 tests)
`pipeline/trainer.py` — Constructs optimizer, scheduler, and loss function internally.
**Fix:** Accept optional `criterion`, `optimizer`, `scheduler` in constructor. Build internally only when not provided.

```python
class SegmentationTrainer:
    def __init__(
        self,
        model: ModelAdapter,
        config: TrainingConfig,
        criterion: nn.Module | None = None,  # NEW: inject or build
        optimizer: Optimizer | None = None,   # NEW
        ...
    ) -> None:
        self.criterion = criterion or build_loss_function(config.loss_name)
```

### R5.13: BentoML accepts ModelAdapter (~2 tests)
`serving/bento_service.py` — Service class should accept `ModelAdapter` as constructor argument instead of calling `bentoml.pytorch.load_model()` directly.

### R5.17: Data loader transform injection (~2 tests)
`data/loader.py` — `build_train_loader()` calls `build_train_transforms(config)` internally. Accept optional `transforms` parameter for testing/customization.

### Protocol-based abstractions (architecture improvement)
Define `Tracker` protocol to decouple `pipeline/trainer.py` from `observability/tracking.py`:
```python
class Tracker(Protocol):
    def log_metrics(self, metrics: dict[str, float], step: int) -> None: ...
    def log_model_info(self, model: Any) -> None: ...
```

## Acceptance Criteria
- [ ] Trainer accepts optional injected dependencies
- [ ] BentoML service accepts ModelAdapter
- [ ] Data loader accepts optional transforms
- [ ] Tracker protocol defined, trainer uses it
- [ ] ~8 new tests added
- [ ] Backwards compatible (existing code unchanged)
